src/modules/wkdLocate.js

```diff
@@ -0,0 +1,240 @@
+ *
+ * Code based on openpgp.js src/util.js rev. 22c66c1
+ * under the terms of the GNU Lesser General Public License Version 3
+ *
+ * @param {String}   email  The canonicalized RFC822 addr spec.
+ *
+ * @returns {String} The WKD URL according to draft-koch-openpgp-webkey-service-06.
+ */
+function buildWKDUrl(email) {
+  const [, localPart, domain] = /(.*)@(.*)/.exec(email);
+
+  if (!localPart || !domain) {
+    throw new Error(`WKD: failed to parse: ${email}`);
+  }
+  const shasum = crypto.createHash('sha1');
+  shasum.update(localPart.toLowerCase());
+  const digest = shasum.digest();
+  const localEncoded = openpgp.util.encodeZBase32(digest);
+
+  return `https://${domain}/.well-known/openpgpkey/hu/${localEncoded}`;
+}
+
+/** Convert a promise into a promise with a timeout.
+  *
+  * @param {Number}    ms       The timeout in milliseconds.
+  * @param {Promise}   promise  The promise to wrap.
+  *
+  * @returns {Promise} A promise with a timeout.
+  **/
+function timeout(ms, promise) {
+  return new Promise(((resolve, reject) => {
+    setTimeout(() => {
+      reject(new Error("WKD: Timeout"));
+    }, ms);
+    promise.then(resolve, reject);
+  }));
+}
+
+/**
+ * A WKD key can contain UserIDs which do not match what
+ * was searched for. This function inspects the response and
+ * filters out UserIDs and keys which do not match, before returning
+ * the keys from the response in armored format to be imported
+ * into the keyring.
+ *
+ * Security: This is a central element for trust by https / WKD through
+ * the domain provider. If accepting any key or any userid provided, there
+ * would be no trust gain as a malicious domain could pollute the keyring
+ * or include userids for other domains.  The validity check and onlyOne
+ * API make this function more convenient to use for simple implementations which
+ * do not plan to use WKD for Key rollover but have no security purpose.
+ *
+ * @param {ByteArray} data          Binary data that might contain keys.
+ * @param {String}    email         The email address for the userids.
+ * @param {Boolean}   onlyOne       Return only the best key. Best means: The newest, valid key.
+ *                                  If no keys are valid it only returns the newest.
+ * @returns {Array[String]}|String  An array of the ASCII armored filtered keys or only one armored key.
+ */
+async function parseKeysForEMail(data, email, onlyOne) {
+  if (!openpgp.util.isUint8Array(data)) {
+    throw new Error(`WKD: parseKeysForEMail invalid data type ${data.constructor.toString()}`);
+  }
+
+  const result = await openpgp.key.read(data);
+  if (result.err) {
+    throw new Error(`WKD: Failed to parse result for '${email}': ${result.err}`);
+  }
+
+  try {
+    const keptArmoredKeys = new Array;
+    let candidate;
+    result.keys.forEach(async key => {
+      const keptUsers = new Array;
+      console.log(`WKD: inspecting: ${key.primaryKey.getFingerprint()}`);
+      key.users.forEach(user => {
```

Comment:
This part should probably go in the key module or somehwere so that we can also use userid filtering for the mvelo keyserver.